### ISSUE TITLE
Crash in DJISDK 4.2.2...

### ISSUE THREAD

**Issue by jgrantr** [dji-sdk/Mobile-SDK-iOS#131]
Yep, then please delete this template and provide the following infos for us to help investigate the issue:

Not sure what is causing this crash for sure, but it does appear to be related to uploading a mission, perhaps after the upload appears to timeout.  A crash log is attached.

Please let me know if you have any further questions.

DJI iOS SDK version = 4.2.2
DJI Product = Phantom 4 Pro
iOS System version = 10.3.2
iOS device type = iPad mini 4 Cellular
Xcode version = 8.3.3

[2017-07-25_12-43-00.1332_-0600-31fc2875d4441ba744cfe053d01b5d884f2e492d.crash.zip](https://github.com/dji-sdk/Mobile-SDK-iOS/files/1175074/2017-07-25_12-43-00.1332_-0600-31fc2875d4441ba744cfe053d01b5d884f2e492d.crash.zip)

**Comment by XietongLU** [comment]
@jgrantr we start to investigate the crash log. We tried to re-symbolicate the crash log but the stack looks strange for me. Please check did you call downloadMissionWithCompletion: within the completion block of uploadMissionWithCompletion:?

**Comment by jgrantr** [comment]
No, we do NOT call downloadMissionWithCompletion anywhere in our code.  However, we do sometimes call upload mission within the execution context of an upload listener block.  I think that is the case here.  We get an error reported in the upload by the upload listener, and then within our listener block, we call uploadMissionWithCompletion:  I think there are also places where we call uploadMissionWithCompletion: inside the completion block of uploadMissionWithCompletion: also (for example, if the upload times out immediately and doesn't actually start).  I actually got this crash flying in the iOS simulator today, and took a screenshot of the call stack.  Here it is:

<img width="583" alt="screen shot 2017-07-26 at 3 00 44 pm" src="https://user-images.githubusercontent.com/1638694/28656112-45b58bb6-725d-11e7-90b2-7c7eddceb7f2.png">

**Comment by XietongLU** [comment]
@jgrantr thanks for your information. "However, we do sometimes call upload mission within the execution context of an upload listener block" could you show us the code about this? Specifically, which event will trigger the uploadMission?

**Comment by jgrantr** [comment]
Here is the code that starts the upload:

 [djiOperator uploadMissionWithCompletion:^(NSError * _Nullable error) {
	if (error && error.code == DJISDKErrorTimeout) {
		if (retries >= MAX_UPLOAD_RETRIES) {
			[self resolveWithError:error];
		}
		else {
			retries++;
			[self uploadMissionInternal];
		}
	}
	else {
		NSLog(@"%@ is now uploading", _operation);
		self.listenForErrors = YES;
	}
}];

And here is the listener block:

^(DJIWaypointMissionUploadEvent * _Nonnull event) {
	if (event.currentState == DJIWaypointMissionStateReadyToExecute) {
		NSLog(@"(%@) upload finished", task.operation);
		[task resolveWithError:nil];
	}
	else if (task.listenForErrors && event.error) {
		if (task.retries >= MAX_UPLOAD_RETRIES) {
			NSLog(@"(%@) error during upload: %@", task.operation, event.error);
			[task resolveWithError:event.error];
		}
		else {
			task.retries = task.retries + 1;;
			NSLog(@"(%@) upload timed out (from observable), doing retry #%ld", task.operation, (long)task.retries);
			[task uploadMissionInternal];
		}
	}
	else if (event.currentState == DJIWaypointMissionStateUploading) {
		NSLog(@"uploading %@, progress = %ld of %lu", task.operation, (long)event.progress.uploadedWaypointIndex, (unsigned long)event.progress.totalWaypointCount);
	}
}

Each time I've seen this, I see a log message about a timeout (-1003) and then it crashes, and the stack traces are what I've sent and the screenshot I've posted.

**Comment by jgrantr** [comment]
Any luck?  Can we expect this to be fixed in 4.3?

**Comment by athiercelin** [comment]
We are looking into timeout errors associated with Waypoints and Timeline. We hope to fix it all for 4.3

**Comment by XietongLU** [comment]
@jgrantr we've fixed this bug internally. The fix will be included in 4.3.

**Comment by zoemcateedji** [comment]
4.3 has been released - please let us know how this is going for you.
